pkg/storaged/nfs-mounts.py

```diff
@@ -0,0 +1,344 @@
+        if sline == "" or sline[0] == "#":
+            new_lines.append(line)
+        else:
+            fields = list(map(field_unescape, re.split("[ \t]+", sline)))
+            if len(fields) > 0 and ":" in fields[0]:
+                new_fields = modify(fields)
+                if new_fields:
+                    if new_fields == fields:
+                        new_lines.append(line)
+                    else:
+                        new_lines.append(" ".join(map(field_escape, new_fields)))
+            else:
+                new_lines.append(line)
+    new_fields = modify(None)
+    if new_fields:
+        new_lines.append(" ".join(map(field_escape, new_fields)))
+
+    f = open(name + ".tmp", "w")
+    f.write("\n".join(new_lines) + "\n")
+    f.flush()
+    os.fsync(f.fileno())
+    f.close()
+    os.rename(name + ".tmp", name)
+
+fstab = [ ]
+fstab_by_remote = { }
+
+mtab = [ ]
+mtab_by_remote = { }
+
+def process_fstab():
+    global fstab, fstab_by_remote
+    fstab = parse_tab("/etc/fstab")
+    fstab_by_remote = index_tab(fstab)
+
+def process_mtab():
+    global mtab, mtab_by_remote
+    mtab = parse_tab("/etc/mtab")
+    mtab_by_remote = index_tab(mtab)
+
+def find_in_tab(tab_by_remote, fields):
+    for t in tab_by_remote.get(fields[0], [ ]):
+        if t[0] == fields[0] and t[1] == fields[1]:
+            return t
+    return None
+
+def report():
+    data = [ ]
+    for f in fstab:
+        m = find_in_tab(mtab_by_remote, f)
+        data.append({ "fstab": True, "fields": f, "mounted": m is not None })
+    for m in mtab:
+        if not find_in_tab(fstab_by_remote, m):
+            data.append({ "fstab": False, "fields": m, "mounted": True })
+    sys.stdout.write(json.dumps(data) + "\n")
+    sys.stdout.flush()
+
+def monitor():
+    process_mtab()
+    process_fstab()
+    report()
+    fstab_watcher = Watcher("/etc/fstab")
+    mtab_file = open("/proc/mounts", "r")
+    while True:
+        (r, w, x) = select.select([ fstab_watcher.inotify.fd ], [ ], [ mtab_file ])
+        if mtab_file in x:
+            process_mtab()
+            report()
+        if fstab_watcher.inotify.fd in r:
+            fstab_watcher.process()
+            process_fstab()
+            report()
+
+def mkdir_if_necessary(path):
+    if not os.path.isdir(path):
+        os.makedirs(path)
+
+def rmdir_if_empty(path):
+    try:
+        os.rmdir(path)
+    except OSError as e:
+        if e.errno != errno.ENOTEMPTY:
+            raise
+
+def update(entry, new_fields):
+    old_fields = entry["fields"]
+    mkdir_if_necessary(new_fields[1])
+    modify_tab("/etc/fstab", lambda fields: new_fields if fields == old_fields else fields)
+    if entry["mounted"]:
+        if (new_fields[0] == old_fields[0]
+            and new_fields[1] == old_fields[1]
+            and new_fields[2] == old_fields[2]):
+            remount(new_fields)
+        else:
+            try:
+                unmount(entry)
+                rmdir_if_empty(old_fields[1])
```

Comment:
Mount points don't have to be directories - you can also e. g. bind-mount a file onto another file. It is correct in this case to not remove the original file and for `rmdir_if_empty()` to fail, but that `errno.ENOTDIR` failure should be handled gracefully. (Not sure if something else here ensures that mount points are always directories).

Reply:
Good point, I didn't think of that.